lib/coupon_code.rb

```diff
@@ -1,46 +1,77 @@
+# frozen_string_literal: true
+
 require 'coupon_code/version'
 require 'securerandom'
 require 'digest/sha1'
 
 module CouponCode
-  SYMBOL = '0123456789ABCDEFGHJKLMNPQRTUVWXY'
+  SYMBOL = '23456789ABCDEFGHJKLMNPQRTUVWXY'
   PARTS  = 3
-  LENGTH = 4
-
-  def self.generate(options = { parts: PARTS })
-    num_parts = options.delete(:parts)
-    parts = []
-    (1..num_parts).each do |i|
-      part = ''
-      (1...LENGTH).each { part << random_symbol }
-      part << checkdigit_alg_1(part, i)
-      parts << part
+  PART_LENGTH = 4
+  # Separators which will never be placed next to each other.
+  # "Curse Word Prevention" algorithm details can be found here
+  # https://arcticicestudio.github.io/icecore-hashids/api/curse-word-prevention.html
+  # "I" and "S" are removed comparing to original algorithm's list
+  # because there are no same characters in the "SYMBOL" list
+  CURSE_WORD_SEPARATORS = %w[C F H T U].freeze
+
+  class << self
+    def generate(options = {})
+      num_parts = options.fetch(:parts, PARTS)
+      part_length = options.fetch(:part_length, PART_LENGTH)
+      parts = []
+      (1..num_parts).each do |part_index|
+        parts << generate_safe_code_part(part_index, part_length)
+      end
+      parts.join('-')
     end
-    parts.join('-')
-  end
 
-  def self.validate(orig, num_parts = PARTS)
-    code = orig.upcase
-    code.gsub!(/[^#{SYMBOL}]+/, '')
-    parts = code.scan(/[#{SYMBOL}]{#{LENGTH}}/)
-    return if parts.length != num_parts
-    parts.each_with_index do |part, i|
-      data  = part[0...(LENGTH - 1)]
-      check = part[-1]
-      return if check != checkdigit_alg_1(data, i + 1)
+    def validate(orig, num_parts = PARTS, part_length = PART_LENGTH)
+      code = orig.upcase
+      code.gsub!(/[^#{SYMBOL}]+/, '')
+      parts = code.scan(/[#{SYMBOL}]{#{part_length}}/)
+      return if parts.length != num_parts
+
+      parts.each_with_index do |part, i|
+        data  = part[0...(part_length - 1)]
+        check = part[-1]
+        return if check != checkdigit_alg_1(data.split(''), i + 1)
+      end
+      parts.join('-')
     end
```

Comment:
For me `validate` method does too many actions, IMO it should be like, separate predicate method `valid?`, and second method - `standartize` which convert string to uppercase and do other stuff.

Reply:
Originally, my focus was on changing the behavior for the `generation` method. Anyway, I've refactored this method too by your request. In fact, `valid?` is a part of the standardization action, so I've made it `private`. 

It seems that maintainer has stopped all activities for this gem. I agree we can make breaking changes by renaming the method to `standardize` and using the same arguments as for `generate` method. Additionally, it might be beneficial to make options like PARTS and PART_LENGTH configurable not at the method's argument level, but it's out of the scope.